src/grammar/ast.ts

```diff
@@ -876,51 +970,61 @@ function eqParameters(
     );
 }
 
-function eqParameterArrays(
-    arr1: AstStructFieldInitializer[],
-    arr2: AstStructFieldInitializer[],
+function eqFieldValues(
+    arg1: AstStructFieldValue,
+    arg2: AstStructFieldValue,
 ): boolean {
-    if (arr1.length !== arr2.length) {
-        return false;
-    }
-
-    for (let i = 0; i < arr1.length; i++) {
-        if (!eqParameters(arr1[i]!, arr2[i]!)) {
-            return false;
-        }
-    }
-
-    return true;
+    return (
+        eqNames(arg1.field, arg2.field) &&
+        eqExpressions(arg1.initializer, arg2.initializer)
+    );
 }
 
-function eqExpressionArrays(
-    arr1: AstExpression[],
-    arr2: AstExpression[],
+function eqArrays<T>(
+    arr1: T[],
+    arr2: T[],
+    eqElements: (elem1: T, elem2: T) => boolean,
 ): boolean {
     if (arr1.length !== arr2.length) {
         return false;
     }
 
     for (let i = 0; i < arr1.length; i++) {
-        if (!eqExpressions(arr1[i]!, arr2[i]!)) {
+        if (!eqElements(arr1[i]!, arr2[i]!)) {
             return false;
         }
     }
 
     return true;
 }
 
-export function isValue(ast: AstExpression): boolean {
+export function isLiteral(ast: AstExpression): ast is AstLiteral {
+    return checkLiteral(
+        ast,
+        () => true,
+        () => false,
+    );
+}
+
+export function checkLiteral<T>(
+    ast: AstExpression,
+    t: (node: AstLiteral) => T,
+    f: (node: Exclude<AstExpression, AstLiteral>) => T,
```

Comment:
Let's document this either locally or as a recommended pattern to use across the codebase